File: features/miso/loop/cards/location/location.js
const feature_Location = {
  // one position per card per page load. A repaint is not a new card, and
  // the observer below fires on every one of them.
  asked: {},
  sheet: null,
  where: null,
  meta: null,

  // "just now" / "3 min ago" / "2 hours ago" — plain words, no clock
  since(t) {
    const ms = Date.now() - t;
    if (!t || !isFinite(ms) || ms < 0) return 'just now';
    const m = Math.floor(ms / 60000);
    if (m < 1) return 'just now';
    if (m < 60) return m + ' min ago';
    const h = Math.floor(m / 60);
    if (h < 24) return h + (h === 1 ? ' hour ago' : ' hours ago');
    const d = Math.floor(h / 24);
    return d + (d === 1 ? ' day ago' : ' days ago');
  },

  // the single geolocation call. Denied, unavailable, timed out or absent:
  // nothing is stored and nothing is said — the pill stays dim, and a tap
  // (`again`) is the one thing that gets a second prompt.
  ask(id, again) {
    if (!id) return;
    if (this.asked[id] && !again) return;
    this.asked[id] = true;
    if (typeof navigator === 'undefined') return;
    const geo = navigator.geolocation;
    if (!geo || typeof geo.getCurrentPosition !== 'function') return;
    try {
      geo.getCurrentPosition((p) => {
        const c = (p && p.coords) || {};
        if (typeof c.latitude !== 'number' || typeof c.longitude !== 'number') return;
        if (typeof feature_Loop === 'undefined' || !feature_Loop.state) return;
        feature_Loop.send({ type: 'CardPlace', data: {
          id,
          lat: c.latitude,
          lon: c.longitude,
          acc: typeof c.accuracy === 'number' ? c.accuracy : 0,
          t: Date.now() } });
      }, () => {}, { enableHighAccuracy: false, timeout: 10000, maximumAge: 300000 });
    } catch (e) {
      /* an API that throws is an API that is not there */
    }
  },

  // the card page appearing is what this node reacts to, read from the DOM
  // rather than by wrapping feature_Loop.apply — that idiom races and orphans
  // other fragments' wrappers (notes.md, "the apply-wrapper race").
  look() {
    const pill = document.querySelector('.card-page .card-place');
    if (!pill) return;
    if (!pill.classList.contains('dim')) return;
    this.ask(pill.getAttribute('data-card'), false);
  },

  // the placeholder view: everything the block knows, straight off the pill's
  // data attributes, so the sheet never reads the store a second time
  show(pill) {
    if (!this.sheet || !pill) return;
    const lat = Number(pill.getAttribute('data-lat'));
    const lon = Number(pill.getAttribute('data-lon'));
    if (!isFinite(lat) || !isFinite(lon)) return;
    const acc = Math.round(Number(pill.getAttribute('data-acc')) || 0);
    const t = Number(pill.getAttribute('data-t')) || 0;
    this.where.textContent = lat.toFixed(4) + ', ' + lon.toFixed(4);
    this.meta.textContent = (acc > 0 ? '±' + acc + 'm · ' : '') + this.since(t);
    this.sheet.classList.add('show');
  },

  hide() {
    if (this.sheet) this.sheet.classList.remove('show');
  },
};

{
  // furniture made at load and living OUTSIDE #app, so a repaint of the
  // loop's html while the sheet is open cannot take it away — the #cardToast
  // and #frameSheet precedent.
  const fm_placeSheet = document.createElement('div');
  fm_placeSheet.id = 'placeSheet';

  const fm_placeBox = document.createElement('div');
  fm_placeBox.id = 'placeBox';
  const fm_placeWhere = document.createElement('div');
  fm_placeWhere.id = 'placeWhere';
  const fm_placeMeta = document.createElement('div');
  fm_placeMeta.id = 'placeMeta';
  const fm_placeClose = document.createElement('button');
  fm_placeClose.id = 'placeClose';
  fm_placeClose.type = 'button';
  fm_placeClose.textContent = 'close';
  fm_placeBox.appendChild(fm_placeWhere);
  fm_placeBox.appendChild(fm_placeMeta);
  fm_placeBox.appendChild(fm_placeClose);
  fm_placeSheet.appendChild(fm_placeBox);
  document.body.appendChild(fm_placeSheet);

  feature_Location.sheet = fm_placeSheet;
  feature_Location.where = fm_placeWhere;
  feature_Location.meta = fm_placeMeta;

  fm_placeClose.addEventListener('click', (e) => {
    e.preventDefault();
    feature_Location.hide();
  });
  // the ground closes it too; the box itself does not
  fm_placeSheet.addEventListener('click', (e) => {
    if (e.target === fm_placeSheet) feature_Location.hide();
  });

  // the pill carries no data-ev, so the loop's own delegated click never
  // fires for it. A dim pill asks again; a lit one opens the place.
  document.addEventListener('click', (e) => {
    if (!e.target || !e.target.closest) return;
    const pill = e.target.closest('.card-place');
    if (!pill) return;
    if (pill.classList.contains('dim')) {
      feature_Location.ask(pill.getAttribute('data-card'), true);
      return;
    }
    feature_Location.show(pill);
  });

  const fm_placeWatch = setInterval(() => {
    const app = document.getElementById('app');
    if (!app) return;
    clearInterval(fm_placeWatch);
    const look = () => feature_Location.look();
    new MutationObserver(look).observe(app, { childList: true, subtree: true });
    look();
  }, 100);
}

File: features/miso/loop/cards/location/location.rs
struct feature_Location;
impl feature_Location {
    // ---- the datum -------------------------------------------------------
    // a card's place is one block in its own body: {kind:"location", lat,
    // lon, acc, t, source}. /cards' page renderer draws nothing for a kind it
    // does not know, so the block is invisible to everything but this node.
    // A garbage block reads the same as no block at all — null.

    fn card_place_of(card: String) -> serde_json::Value {
        let c: serde_json::Value = serde_json::from_str(&card)
            .unwrap_or(serde_json::Value::Null);
        let empty: Vec<serde_json::Value> = Vec::new();
        for b in c["blocks"].as_array().unwrap_or(&empty) {
            if b["kind"].as_str().unwrap_or("") != "location" {
                continue;
            }
            let lat = b["lat"].as_f64().unwrap_or(f64::NAN);
            let lon = b["lon"].as_f64().unwrap_or(f64::NAN);
            if card_place_sound(lat, lon) {
                return b.clone();
            }
            return serde_json::Value::Null;
        }
        serde_json::Value::Null
    }

    // the one test a coordinate has to pass, on the way in and on the way
    // out: a place that is not a place never enters the world and never
    // reaches the screen.
    fn card_place_sound(lat: f64, lon: f64) -> bool {
        lat.is_finite() && lon.is_finite()
            && lat >= -90.0 && lat <= 90.0
            && lon >= -180.0 && lon <= 180.0
    }

    // ---- the event -------------------------------------------------------
    // CardPlace {id, lat, lon, acc, t}: one card carries at most one place,
    // so a second fix replaces the first. Read and written through /cards'
    // own cards_read / cards_write, so the var's address stays in one place
    // and cards.rs is never edited.

    fn update(state: String, event: String) -> String {
        let state = existing.update(state, event.clone());
        let e: serde_json::Value = serde_json::from_str(&event)
            .unwrap_or(serde_json::Value::Null);
        if e["type"].as_str().unwrap_or("") != "CardPlace" {
            return state;
        }
        let id = e["data"]["id"].as_str().unwrap_or("").to_string();
        let lat = e["data"]["lat"].as_f64().unwrap_or(f64::NAN);
        let lon = e["data"]["lon"].as_f64().unwrap_or(f64::NAN);
        if id.is_empty() || !card_place_sound(lat, lon) {
            return state;
        }
        let acc = e["data"]["acc"].as_f64().unwrap_or(0.0);
        let acc = if acc.is_finite() && acc >= 0.0 { acc } else { 0.0 };
        let now = e["data"]["t"].as_u64().unwrap_or(0);
        let mut list: serde_json::Value = serde_json::from_str(&cards_read())
            .unwrap_or(serde_json::json!([]));
        if !list.is_array() {
            return state;
        }
        let block = serde_json::json!({
            "kind": "location",
            "lat": lat,
            "lon": lon,
            "acc": acc,
            "t": now,
            "source": "device"
        });
        let mut changed = false;
        for c in list.as_array_mut().expect("cards is an array").iter_mut() {
            if c["id"].as_str().unwrap_or("") != id {
                continue;
            }
            if !c["blocks"].is_array() {
                continue;
            }
            let blocks = c["blocks"].as_array_mut().expect("blocks is an array");
            let mut at = blocks.len();
            for i in 0..blocks.len() {
                if blocks[i]["kind"].as_str().unwrap_or("") == "location" {
                    at = i;
                    break;
                }
            }
            if at < blocks.len() {
                blocks[at] = block.clone();
            } else {
                blocks.push(block.clone());
            }
            c["edited"] = serde_json::json!(now);
            changed = true;
        }
        if changed {
            cards_write(list.to_string());
        }
        state
    }

    // ---- the pill --------------------------------------------------------
    // spliced in before the page's LAST closing div: inside the card's own
    // scrolling box, after the blocks, and before whatever me_under puts
    // under the card. No data-ev on it, so the loop's delegated click never
    // fires and a tap cannot repaint #app out from under itself.

    fn card_page_html(card: String) -> String {
        let html = existing.card_page_html(card.clone());
        let c: serde_json::Value = serde_json::from_str(&card)
            .unwrap_or(serde_json::Value::Null);
        let id = card_esc(c["id"].as_str().unwrap_or("").to_string());
        let at = card_place_of(card.clone());
        let pill = if at.is_null() {
            format!(
                "<span class=\"card-place dim\" data-card=\"{}\">map location</span>",
                id)
        } else {
            format!(
                "<span class=\"card-place\" data-card=\"{}\" data-lat=\"{:.5}\" data-lon=\"{:.5}\" data-acc=\"{}\" data-t=\"{}\">map location</span>",
                id,
                at["lat"].as_f64().unwrap_or(0.0),
                at["lon"].as_f64().unwrap_or(0.0),
                at["acc"].as_f64().unwrap_or(0.0).round() as i64,
                at["t"].as_u64().unwrap_or(0))
        };
        match html.rfind("</div>") {
            Some(i) => format!("{}{}{}", &html[..i], pill, &html[i..]),
            None => format!("{}{}", html, pill),
        }
    }
}
